doc:ipsec:kvstore: explicit limitations that could lead to staling XFRM states and no connectivity

smagnani96 opened the pull request:
IPSec relies on XFRM states and policies to ensure encrypted pod-to-pod connectivity. In the XFRM state:
    
* seq is the incoming sequence number counter. It is used by the kernel
      to track and validate received packets. When a packet arrives, its
      sequence number is compared to the expected range to detect replays
      or out-of-order packets. Together this the oseq value, it helps implement
      the anti-replay window to prevent attackers from resending previously captured packets.
* oseq is the outgoing sequence number counter. It is used when sending
      packets protected by IPsec. Each outbound IPsec packet is assigned an
      incrementing oseq value. The oseq ensures unique sequence numbers for
      each packet, which the receiver uses to validate the order and detect replays.
    
In a SA between two nodes A and B, the seq/oseq values in the XFRM state A must match the oseq/seq values in node B, and vice versa. If that is not the case, users would experience the `XfrmInStateProtoError` error, with no IPSec connectivity between the two nodes.
    
We noticed that a Cilium user might end up in this situation in both the following cases, as stated in the doc changes:
    
1. KVStore Mode (e.g., etcd): When a Cilium agent connects too late to the
       newly created KVStore, it may miss the node delete and create events
       for entries that were restored or reinitialized. This results in staling
       XFRM state, causing permanent network disruption. Please do note that
       operations such as etcd removal and recreation are not supported.
2. CRD Mode: a similar issue may occur when a CiliumNode resource is deleted
       and the Cilium agent DaemonSet is restarted. While other agents will
       recreate fresh XFRM states for the new CiliumNode, the restarted agent
       may continue to hold obsolete XFRM states referencing all peer nodes.
    
The identified mitigation strategy for these scenario is a proper IPSec key rotation, which would cause all the states to be consistently recreated in all Cilium agents.

joestringer commented on Documentation/kvstore.rst:
```diff
@@ -4,6 +4,12 @@ Key-Value Store
 
+.. note::
+
+    Cilium does not support removing and recreating a kvstore that is currently in use.
```
Hmm. Since the beginning, we often stated that the source of truth for kvstore state was in the agents, and therefore if you deleted the content of the KVStore, all agents would recreate the state and recover.

It sounds like at some point the IPSec feature specifically changed this assumption. Is that fair to say?

joestringer replied:
Since we never stated that removing kvstore state is OK in the docs, and we probably aren't confident enough to make such a statement now, maybe the best approach is to word the statement roughly with the following points:

- Don't remove kvstore state manually unless you know exactly what you're doing.
- If you manually remove kvstore state when transparent encryption is enabled then that may cause permanent connectivity disruption. (for pods managed by Cilium? for nodes?)

smagnani96 replied:
> It sounds like at some point the IPSec feature specifically changed this assumption. Is that fair to say?

Yes, I'd say so. We do not guarantee that **ALL** XFRM states are recreated and in-sync.
I opened an issue to track this https://github.com/cilium/cilium/issues/39720.

joestringer commented on Documentation/kvstore.rst:
```diff
@@ -4,6 +4,12 @@ Key-Value Store
 Cilium uses an external key-value store to exchange information across multiple
 Cilium instances:
 
+.. note::
```
Rather than creating a `.. note::` here, I would suggest creating a new section `Caveats and Limitations` above `Debugging`, and put the text under that section with no highlights.

joestringer commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -328,6 +328,21 @@ errors.
 
+  * KVStore Mode (e.g., etcd): XFRM state inconsistencies can occur if the KVStore
+    is deleted and subsequently recreated — a scenario that is explicitly unsupported
+    by Cilium. When a Cilium agent connects too late to the newly created KVStore,
```
As a general rule, "support" can imply a relationship between users and the project that doesn't exist. As such I typically discourage using that wording. It's often better to talk about facts "If you do X, Y will happen". There's a range of examples to encourage the right tone in [this docs style guide section](https://docs.cilium.io/en/stable/contributing/docs/docsstyle/#common-pitfalls).

```suggestion
    is deleted and subsequently recreated. When a Cilium agent connects too late to the newly created KVStore,
```

I think that the admonishment in `kvstore.rst` should be sufficient to cover this case. Manually editing Cilium kvstore content should be an exceedingly rare situation, and if this section already says "here are the consequences" then I don't think we need to further emphasize the point.

joestringer commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -328,6 +328,21 @@ errors.
+  * KVStore Mode (e.g., etcd): XFRM state inconsistencies can occur if the KVStore
+    is deleted and subsequently recreated — a scenario that is explicitly unsupported
+    by Cilium. When a Cilium agent connects too late to the newly created KVStore,
+    it may miss the node delete and create events for entries that were restored or
```
```suggestion
    Cilium may miss the node delete and create events for entries that were restored or
```

joestringer commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -328,6 +328,21 @@ errors.
    from the anti-reply oseq on the destination node. This can be fixed by
    properly performing a new key rotation.
 
+  * KVStore Mode (e.g., etcd): XFRM state inconsistencies can occur if the KVStore
```
As a matter of structure, all of the other paragraphs in this section seem to be about specific errors returned by Xfrm. However, these two new paragraphs are general statements, and they are located in the middle of the section with specific error code names above and below these paragraphs. Probably it makes more sense to move these paragraphs out to either (a) the end of this section or (b) as part of a new subsection.

joestringer reviewed (changes requested):
I provided a few general tips to follow the docs-structure style that we are aiming for in the Cilium docs. I will defer on the technical content to the relevant code owner. See the comments below for more details.

joestringer reviewed (approved):
Looks good to me for @cilium/docs-structure .

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -387,6 +387,47 @@ errors.
+pods managed by Cilium. You can identify this problem by monitoring the
+``XfrmInStateProtoError`` metric; an increasing value indicates an issue.
+This error occurs because of mismatched ``seq/oseq`` values between the new and
+old XFRM states on two CiliumNodes.
```
That's not the only explanation for this error unfortunately. And I suspect this is too much details as it's hardly actionnable for users.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -387,6 +387,47 @@ errors.
+old XFRM states on two CiliumNodes.
+
+To validate this, inspect the XFRM states directly from a Cilium agent pod
+``kubectl -n kube-system exec -ti ds/cilium -- ip xfrm state``
```
Uh, do we really expect users to dig into XFRM states and match them between different nodes to figure this out?

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -387,6 +387,47 @@ errors.
+----------
+
+To prevent and resolve stale XFRM states, perform a proper key rotation.
+This action ensures new consistent and valid XFRM states across all your nodes.
```
I'd drop "proper" (or is there such a thing as an improper key rotation?). Let's also link to the appropriate section.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -387,6 +387,47 @@ errors.
+XFRM State Staling in Cilium
+============================
+
+XFRM state inconsistencies can disrupt connectivity for pods managed by Cilium.
```
Would it be more appropriate to say that control plane disruptions can lead to connectivity issues due to IPsec anti-replay counters getting out of sync? I'd like to make a bit clearer that this only happens in extreme situations.

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -387,6 +387,47 @@ errors.
+    the agent to miss crucial node delete and create events, leading Cilium to
+    retain outdated XFRM states for those nodes.
+
+  * CRD Mode: A similar problem occurs when you delete a CiliumNode resource and
```
when -> if

It's not as if we expect users to ever have to legitimately do this.

pchaigno reviewed (commented):
I think this needs to be a reworded a bit. As it stands, it seems likely to worry unaffected users and to be unactionnable for affected users.

smagnani96 committed:
doc:ipsec: explicit limitation with stale XFRM states in CRD/KVStore

IPSec relies on XFRM states and policies to ensure encrypted pod-to-pod
connectivity. In the XFRM state:

* seq is the incoming sequence number counter. It is used by the kernel
  to track and validate received packets. When a packet arrives, its
  sequence number is compared to the expected range to detect replays
  or out-of-order packets. Together this the oseq value, it helps implement
  the anti-replay window to prevent attackers from resending previously captured packets.
* oseq is the outgoing sequence number counter. It is used when sending
  packets protected by IPsec. Each outbound IPsec packet is assigned an
  incrementing oseq value. The oseq ensures unique sequence numbers for
  each packet, which the receiver uses to validate the order and detect replays.

In a SA between two nodes A and B, the seq/oseq values in the XFRM state
A must match the oseq/seq values in node B, and vice versa. If that is not
the case, users would experience the `XfrmInStateProtoError` error, with
no IPSec connectivity between the two nodes.

We noticed that a Cilium user might end up in this situation in both the
following cases, as stated in the doc changes:

1. KVStore Mode (e.g., etcd): if a Cilium agent connects too late to the
   newly created KVStore, it may miss the node delete and create events
   for entries that were restored or reinitialized. This results in staling
   XFRM state, causing permanent network disruption.
2. KVStore Mode: if a Cilium agent is down for prolonged time, the
   corresponding node entry in the kvstore will be deleted due to lease
   expiration (15m), resulting in stale XFRM states.
3. CRD Mode: a similar issue may occur when a CiliumNode resource is deleted
   and the Cilium agent DaemonSet is restarted. While other agents will
   recreate fresh XFRM states for the new CiliumNode, the restarted agent
   may continue to hold obsolete XFRM states referencing all peer nodes.

The identified mitigation strategy for these scenario is an IPSec
key rotation, which would cause all the states to be consistently
recreated in all Cilium agents.

Signed-off-by: Simone Magnani <simone.magnani@isovalent.com>

pchaigno commented on Documentation/security/network/encryption-ipsec.rst:
```diff
@@ -398,6 +398,42 @@ errors.
+Mitigation
+----------
+
+To prevent and resolve stale XFRM states, perform a key rotation (see
```
It's not a preventive action, is it? How about "To restore connectivity in those cases, perform..."

pchaigno reviewed (approved):
That's a lot better IMO! One small suggestion below.
